File: 백준/Gold/2922. 즐거운 단어/즐거운 단어.cc
#include <iostream>
#include <vector>
using namespace std;

string s;	   // 입력
vector<int> v; // 밑줄이 있는 인덱스를 저장하는 리스트
long long ans = 0;

bool isMo(char c)
{
	return c == 'A' || c == 'E' || c == 'I' || c == 'O' || c == 'U';
}

bool check(string t, bool isL)
{
	if (!isL)
	{
		return false;
	}

	for (int i = 0; i < t.length() - 2; i++)
	{
		if (isMo(t[i]) && isMo(t[i + 1]) && isMo(t[i + 2]))
		{
			return false;
		}
		if (!isMo(t[i]) && !isMo(t[i + 1]) && !isMo(t[i + 2]))
		{
			return false;
		}
	}
	return true;
}

void go(int idx, bool isL, long long cnt, string t)
{
	if (idx == v.size())
	{
		if (check(t, isL))
		{
			ans += cnt;
		}
		return;
	}

	t[v[idx]] = 'L';
	go(idx + 1, true, cnt, t); // l
	t[v[idx]] = 'A';
	go(idx + 1, isL, cnt == 0 ? 5 : cnt * 5, t); // 모음
	t[v[idx]] = 'B';
	go(idx + 1, isL, cnt == 0 ? 20 : cnt * 20, t); // 자음
}

int main()
{
	cin >> s;
	bool isL = false;
	for (int i = 0; i < s.length(); i++)
	{
		if (s[i] == '_')
		{
			v.push_back(i);
		}
		if (s[i] == 'L')
		{
			isL = true;
		}
	}

	go(0, isL, 0, s);

	cout << ans;
}

File: 백준/Gold/2922. 즐거운 단어/즐거운 단어.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {

	static char[] s;
	static long ans;

	public static void main(String[] args) throws Exception {
		var br = new BufferedReader(new InputStreamReader(System.in));

		String inp = br.readLine();
		s = new char[inp.length()];
		for (int i = 0; i < inp.length(); i++) {
			s[i] = inp.charAt(i);
		}
		go(0);
		System.out.println(ans);
	}

	static void go(int idx) {
		if (idx == s.length) {
			if (!check()) {
				return;
			}
			long cnt = 1;
			for (int i = 0; i < s.length; i++) {
				if (s[i] == '*') {
					cnt *= 20;
				} else if (s[i] == '+') {
					cnt *= 5;
				}
			}
			// for(int i=0;i<s.length;i++){
			// 	System.out.print(s[i]);
			// }
			// System.out.println();
			// // 5 + 5*27 + 5 + 5*27 + 5*5 + 27*5 + 27*5 + 5*5 + 5*5
			ans += cnt;
			return;
		}

		if (s[idx] == '_') {
			s[idx] = 'L';
			go(idx + 1);

			// 자음
			s[idx] = '*';
			go(idx + 1);

			// 모음
			s[idx] = '+';
			go(idx + 1);
			s[idx] = '_';
		} else {
			go(idx + 1);
		}
	}

	static boolean check() {
		boolean l = false;
		for (int i = 0; i < s.length; i++) {
			if (s[i] == 'L') {
				l = true;
			}
			if (i >= 2) {
				if (isJ(s[i - 2]) && isJ(s[i - 1]) && isJ(s[i])) {
					return false;
				}
				if (!isJ(s[i - 2]) && !isJ(s[i - 1]) && !isJ(s[i])) {
					return false;
				}
			}
		}
		return l;
	}

	static boolean isJ(char c) {
		return !(c == 'A' || c == 'E' || c == 'I' || c == 'O' || c == 'U' || c == '+');
	}
}